document.py

```diff
@@ -18,93 +18,153 @@ def __init__(self, sizeX: int, sizeY: int) -> None:
+            newPos2Code = self.GetCellCode(grabbedNewPos2[0], grabbedNewPos2[1])
+            newPos1Cell = self.GetCell(grabbedNewPos1[0], grabbedNewPos1[1])
+            newPos2Cell = self.GetCell(grabbedNewPos2[0], grabbedNewPos2[1])
+
+            # new positions must be empty or occupied by the same table or agent
+            if (newPos1Code != CellCodes.Empty and newPos1Cell not in [grabbedTable, agent]) \
+                or (newPos2Code != CellCodes.Empty and newPos2Cell not in [grabbedTable, agent]):
+                objMoveSuccessful = False
+            else:
+                # move object
+                objMoveSuccessful = True
+                grabbedTable.x1 = grabbedNewPos1[0]
+                grabbedTable.y1 = grabbedNewPos1[1]
+                grabbedTable.x2 = grabbedNewPos2[0]
+                grabbedTable.y2 = grabbedNewPos2[1]
+                self.grid[grabbedPos1[0]][grabbedPos1[1]] = None
+                self.grid[grabbedPos2[0]][grabbedPos2[1]] = None
+                self.grid[grabbedNewPos1[0]][grabbedNewPos1[1]] = grabbedTable
+                self.grid[grabbedNewPos2[0]][grabbedNewPos2[1]] = grabbedTable
+
+        if objMoveSuccessful:
+            if self.GetCell(agent.x, agent.y) != grabbedObj:    # issue when going opposite direction to grab direction
+                self.grid[agent.x][agent.y] = None
+            agent.x = targetPosition[0]
+            agent.y = targetPosition[1]
+            self.grid[agent.x][agent.y] = agent
+
+            self.validate()
+            return True
+        else:
+            return False
+
+    def moveActionSoloAgent(self, agent: Agent, direction: str) -> bool:
+        assert(agent.grab is None)
+
+        agentPos = np.array([agent.x, agent.y])
+
+        targetPosition = agentPos + Action.directionToVector(direction)
+        
+        targetCode = self.GetCellCode(targetPosition[0], targetPosition[1])
+
+        # target cell must be empty
+        if targetCode != CellCodes.Empty:   # TODO stronger check, what if many agents want to move to the same cell in the same step
+            return False
+
+        self.grid[agent.x][agent.y] = None
+        agent.x = targetPosition[0]
+        agent.y = targetPosition[1]
+        self.grid[agent.x][agent.y] = agent
+
+        return True
+
+    def grabActionAgent(self, agent: Agent, direction: str) -> bool:
+        if(agent.grab is not None):
+            return False
+        agentPos = np.array([agent.x, agent.y])
+
+        targetPosition = agentPos + Action.directionToVector(direction)
+
+        # check if there's furniture there
+        target = self.GetCell(targetPosition[0], targetPosition[1])
+        targetCode = self.GetCellCode(targetPosition[0], targetPosition[1])
+        if not CellCodes.IsGrabbable(targetCode):
+            return False
+
+        # TODO: consider legal table-carrying configurations (can you only carry by holding short ends?)
+        agent.grab = direction
```

Comment:
we might need to add a grab state to tables/chairs as well to keep track of how many agents are holding them and where they're trying to pull.

Reply:
ok, good idea. It will require more care when grabbing / releasing, but will be useful. Ill add it to validation as well